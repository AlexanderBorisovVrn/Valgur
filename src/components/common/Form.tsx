import { FC } from "react";
import {
  Box,
  Typography,
  FormControl,
  FormHelperText,
  TextareaAutosize,
  Stack,
  Select,
  MenuItem,
  Button,
  TextField,
} from "@pankod/refine-mui";
import { IForm } from "interfaces/components";
import CustomButton from "./CustomButton";
import { PreviewImage } from "./PreviewImage";

const Form: FC<IForm> = ({
  type,
  formLoading,
  handleImageChange,
  handleSubmit,
  onFinishHandler,
  propertyImage,
  register,
}) => {
  return (
    <Box>
      <Typography fontSize={25} fontWeight={700} color="#11142d">
        {type} a Property
      </Typography>
      <Box mt={2.5} borderRadius="16px" padding="20px" bgcolor="#fcfcfc">
        <form
          onSubmit={handleSubmit(onFinishHandler)}
          style={{
            width: "100%",
            display: "flex",
            flexDirection: "column",
            gap: "20px",
          }}
        >
          <FormControl>
            <FormHelperText
              sx={{
                fontWeight: 500,
                margin: "10px 0",
                fontSize: 16,
                color: "#11124d",
              }}
            >
              Enter property name
            </FormHelperText>
            <TextField
              fullWidth
              required
              id="outlined-basic"
              color="info"
              variant="outlined"
              autoComplete="hello"
              {...register("title", {
                required: true,
              })}
            />
          </FormControl>
          <FormControl>
            <FormHelperText
              sx={{
                fontWeight: 500,
                margin: "10px 0",
                fontSize: 16,
                color: "#11124d",
              }}
            >
              Enter Description
            </FormHelperText>
            <TextareaAutosize
              minRows={6}
              required
              placeholder="Write a description"
              style={{
                width: "100%",
                background: "transparent",
                fontSize: "16px",
                borderColor: "rgba(0,0,0,0.23)",
                borderRadius: 6,
                padding: 10,
                color: "#919191",
              }}
              {...register("description", {
                required: true,
              })}
            />
          </FormControl>
          <Stack direction="row" gap={4}>
            <FormControl sx={{ flex: 1 }}>
              <FormHelperText
                sx={{
                  fontWeight: 500,
                  margin: "10px 0",
                  fontSize: 16,
                  color: "#11142d",
                }}
              >
                Select Property Type
              </FormHelperText>
              <Select
                variant="outlined"
                color="info"
                required
                displayEmpty
                inputProps={{
                  "aria-label": "Without label",
                }}
                defaultValue="apartment"
                {...register("propertyType", {
                  required: true,
                })}
              >
                <MenuItem value="apartment">Apartment</MenuItem>
                <MenuItem value="villa">Villa</MenuItem>
                <MenuItem value="farmhouse">Farmhouse</MenuItem>
                <MenuItem value="condos">Condos</MenuItem>
                <MenuItem value="townhouse">Townhouse</MenuItem>
                <MenuItem value="duplex">Duplex</MenuItem>
                <MenuItem value="studio">Studio</MenuItem>
                <MenuItem value="chalet">Chalet</MenuItem>
              </Select>
            </FormControl>

            <FormControl>
              <FormHelperText
                sx={{
                  fontWeight: 500,
                  margin: "10px 0",
                  fontSize: 16,
                  color: "#11124d",
                }}
              >
                Enter property price
              </FormHelperText>
              <TextField
                fullWidth
                required
                type="number"
                id="outlined-basic"
                color="info"
                variant="outlined"
                autoComplete="hello"
                {...register("price", {
                  required: true,
                })}
              />
            </FormControl>
          </Stack>
          <FormControl>
            <FormHelperText
              sx={{
                fontWeight: 500,
                margin: "10px 0",
                fontSize: 16,
                color: "#11124d",
              }}
            >
              Enter Location
            </FormHelperText>
            <TextField
              fullWidth
              required
              id="outlined-basic"
              color="info"
              variant="outlined"
              autoComplete="hello"
              {...register("location", {
                required: true,
              })}
            />
          </FormControl>

          <Stack direction="column" gap={1} justifyContent="center" mb={2}>
            <Stack direction="row" gap={2} alignItems="center">
              <Typography
                color="#11142d"
                fontSize={16}
                fontWeight={500}
                my="10px"
              >
                Property Photo
              </Typography>
              <Button
                component="label"
                sx={{
                  width: "fit-content",
                  color: "#2ed480",
                  textTransform: "capitalize",
                  fontSize: 16,
                }}
              >
                Upload
                <input
                  hidden
                  accept="image/*"
                  type="file"
                  onChange={(e) => {
                    //@ts-ignore
                    handleImageChange(e.target.files[0]);
                  }}
                />
              </Button>
            </Stack>
            {propertyImage?.name ? (
              <Stack
                direction="row"
                gap={1}
                sx={{
                  padding: "5px 10px",
                }}
              >
                <PreviewImage
                  cb={handleImageChange}
                  url={propertyImage.url}
                  name={propertyImage.name}
                />
              </Stack>
            ) : null}
          </Stack>

          <CustomButton
            color="#fcfcfc"
            title={formLoading ? "Submitting..." : "Submit"}
            type="submit"
            backgroundColor="#475be8"
          />
        </form>
      </Box>
    </Box>
  );
};

export default Form;
