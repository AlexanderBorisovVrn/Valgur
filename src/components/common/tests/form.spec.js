import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Form from '../Form';
import { FormProvider } from "@pankod/refine-react-hook-form";
import React from 'react';

describe('Form component', () => {
    const mockOnFinishHandler = jest.fn();
    const Wrapper = (props) => {
        return (
            <FormProvider register={() => { }} handleSubmit={() => { }}>
                {props.children}
            </FormProvider>
        );
    };


    test('should render the basic fields', () => {

        render(
            <Wrapper>
                <Form
                    type="Test"
                    formLoading={false}
                    handleImageChange={jest.fn()}
                    onFinishHandler={mockOnFinishHandler}
                    propertyImage={null}

                />
            </Wrapper>
        );

        expect(screen.getByTestId('property-name')).toBeInTheDocument();

        expect(screen.getByTestId('property-description')).toBeInTheDocument();

        expect(screen.getByTestId('property-type')).toBeInTheDocument();

        expect(screen.getByTestId('property-price')).toBeInTheDocument();
        expect(screen.getByTestId('upload-button')).toBeInTheDocument();

    });


    test('upload  files', () => {
        render(
            <Wrapper>
                <Form
                    type="Test"
                    formLoading={false}
                    handleImageChange={jest.fn()}
                    onFinishHandler={mockOnFinishHandler}
                    propertyImage={null}

                />
            </Wrapper>
        );
       
        const file = new File(['file content'], 'image.png', { type: 'image/png' });

        // Нахождение <input> элемента
        const inputElement = screen.getByTestId('upload-input');
      
        // Имитация события изменения файла
        fireEvent.change(inputElement, { target: { files: [file] } });

    })
    test('renders Form correctly', () => {
        const { asFragment } = render(
            <Wrapper>
                <Form
                    type="Test"
                    formLoading={false}
                    handleImageChange={jest.fn()}
                    onFinishHandler={mockOnFinishHandler}
                    propertyImage={null}

                />
            </Wrapper>
        );;
        expect(asFragment()).toMatchSnapshot();
    });

});


