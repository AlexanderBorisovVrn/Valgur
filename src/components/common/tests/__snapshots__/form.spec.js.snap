// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders Form correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-grki9k-MuiTypography-root"
    >
      Test a Property
    </p>
    <div
      class="MuiBox-root css-1e4gsuf"
    >
      <form
        style="width: 100%; display: flex; flex-direction: column; gap: 20px;"
      >
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-si7cvk-MuiFormHelperText-root"
          >
            Enter property name
          </p>
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            data-testid="property-name"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorInfo MuiInputBase-fullWidth MuiInputBase-formControl css-1udmear-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r8:"
                required=""
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <p
            class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-si7cvk-MuiFormHelperText-root"
          >
            Enter Description
          </p>
          <textarea
            data-testid="property-description"
            placeholder="Write a description"
            required=""
            rows="6"
            style="height: 0px; width: 100%; background: transparent; font-size: 16px; border-color: rgba(0,0,0,0.23); border-radius: 6px; padding: 10px; color: rgb(145, 145, 145); overflow: hidden;"
          />
          <textarea
            aria-hidden="true"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%; background: transparent; font-size: 16px; border-color: rgba(0,0,0,0.23); border-radius: 6px; padding: 0px; color: rgb(145, 145, 145);"
            tabindex="-1"
          />
        </div>
        <div
          class="MuiStack-root css-1x8uzb8-MuiStack-root"
        >
          <div
            class="MuiFormControl-root css-11njvxm-MuiFormControl-root"
          >
            <p
              class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled css-d8p0nm-MuiFormHelperText-root"
            >
              Select Property Type
            </p>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorInfo MuiInputBase-formControl  css-1u8u423-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              data-testid="property-type"
              required=""
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                aria-label="none"
                class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                role="button"
                tabindex="0"
              >
                Apartment
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                required=""
                tabindex="-1"
                value="apartment"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
          >
            <p
              class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-si7cvk-MuiFormHelperText-root"
            >
              Enter property price
            </p>
            <div
              aria-label="Price"
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              data-testid="property-price"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorInfo MuiInputBase-fullWidth MuiInputBase-formControl css-1udmear-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  autocomplete="hello"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r9:"
                  required=""
                  type="number"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiAutocomplete-root MuiAutocomplete-fullWidth MuiAutocomplete-hasPopupIcon css-xg9u7b-MuiAutocomplete-root"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
          >
            <p
              class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-si7cvk-MuiFormHelperText-root"
            >
              Enter Location
            </p>
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorInfo MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1gspjga-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-autocomplete="both"
                  aria-expanded="false"
                  aria-invalid="false"
                  autocapitalize="none"
                  autocomplete="off"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                  id="suggest-location"
                  required=""
                  role="combobox"
                  spellcheck="false"
                  type="text"
                  value=""
                />
                <div
                  class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                >
                  <button
                    aria-label="Open"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                    tabindex="-1"
                    title="Open"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="ArrowDropDownIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiStack-root css-1kxrjq8-MuiStack-root"
        >
          <div
            class="MuiStack-root css-12djzut-MuiStack-root"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-u2ni3s-MuiTypography-root"
            >
              Property Photo
            </p>
            <label
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-dwqlmc-MuiButtonBase-root-MuiButton-root"
              data-testid="upload-button"
              role="button"
              tabindex="0"
            >
              Upload
              <input
                accept="image/*"
                data-testid="upload-input"
                hidden=""
                type="file"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </label>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1d8qxke-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Submit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
